File: problems/leetcode/LongestIncreasingPathInAMatrix.java
// https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
class Solution {
    int[][] dir = new int[][] {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    public int longestIncreasingPath(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }

        int M = matrix.length;
        int N = matrix[0].length;

        int[][] memo = new int[M][N];
        int longest = 0;
        for (int i = 0; i < M; i++) {
            for (int j = 0; j < N; j++) {
                longest = Math.max(dfs(matrix, i, j, memo), longest);
            }
        }

        return longest;
    }

    private int dfs(int[][] matrix, int i, int j, int[][] memo) {
        if (memo[i][j] > 0) {
            return memo[i][j];
        }

        for (int[] curDir : dir) {
            int curI = i + curDir[0];
            int curJ = j + curDir[1];

            if (curI < 0 || curI >= matrix.length ||
                curJ < 0 || curJ >= matrix[0].length ||
                matrix[i][j] >= matrix[curI][curJ]) {
                continue;
            }

            memo[i][j] = Math.max(memo[i][j], dfs(matrix, curI, curJ, memo));
        }

        return ++memo[i][j];
    }
}

File: problems/leetcode/longest-increasing-path-in-a-matrix.py
# https://leetcode.com/problems/longest-increasing-path-in-a-matrix/
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        
        def getLongestPath(X, Y):
            if not (0 <= X < len(matrix) and 0 <= Y < len(matrix[0])):
                return 0
            
            if memo[X][Y] == -1:
                memo[X][Y] = 1
                for offsetX, offsetY in directions:
                    nextX, nextY = X + offsetX, Y + offsetY
                    if 0 <= nextX < len(matrix) and 0 <= nextY < len(matrix[0]) and \
                        matrix[nextX][nextY] > matrix[X][Y]:
                        memo[X][Y] = max(memo[X][Y], getLongestPath(nextX, nextY) + 1)
                    
            return memo[X][Y]
        
        
        memo = [[-1] * len(matrix[0]) for _ in range(len(matrix))]
        maxPath = 0
        
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                maxPath = max(maxPath, getLongestPath(i, j))
                
        return maxPath
